app/services/unsubscribe_all_service.rb

```diff
@@ -7,6 +7,10 @@ def initialize(subscriber, reason, **)
   end
 
   def call
-    UnsubscribeService.call(subscriber, subscriber.active_subscriptions, reason)
+    ActiveRecord::Base.transaction do
+      subscriber.active_subscriptions.each do |subscription|
+        subscription.end(reason: reason)
```

Comment:
It feels a bit superfluous to be doing an n+1 to update each individual subscription here, given there isn't really any logic involved in the `end` method that is applicable.

You could do:

```
ended_count= subscriber.active_subscriptions.update_all(ended_at: Time.zone.now, ended_reason: reason)
```

To do it in one query (which would also save the need for a transaction as it would be atomic). You wouldn't need to worry about the check for ended or not since this would only be applying this to subscriptions that weren't ended.

Only thing that'd be missing would be the metrics. You could update `Metrics.unsubscribed` to accept a second parameter of the count.

Reply:
Sure, have done your recommendation and updated the `unsubscribed` method to take a count.